tmf/org.eclipse.tracecompass.tmf.core/src/org/eclipse/tracecompass/tmf/core/trace/TmfTrace.java

```diff
@@ -784,4 +780,53 @@ public void setComplete(boolean isComplete) {
          * reading (traces in an incomplete state)
          */
     }
+
+    private IStatus refreshAnalysisModulesImpl() {
+        Map<String, IAnalysisModule> previousAnalysisModules = new HashMap<>(fAnalysisModules);
+        Map<String, IAnalysisModule> newAnalysisModules = new HashMap<>();
+        MultiStatus status = new MultiStatus(Activator.PLUGIN_ID, IStatus.OK, null, null);
+        /* First modules are initialized */
+        Map<String, IAnalysisModuleHelper> modules = TmfAnalysisManager.getAnalysisModules(this.getClass());
+        for (IAnalysisModuleHelper helper : modules.values()) {
+            try {
+                IAnalysisModule module = helper.newModule(this);
+                if (module == null) {
+                    continue;
+                }
+                newAnalysisModules.put(module.getId(), module);
+            } catch (TmfAnalysisException e) {
+                status.add(new Status(IStatus.WARNING, Activator.PLUGIN_ID, e.getMessage()));
+            }
+        }
+        Map<String, IAnalysisModule> oldAnalysisModules = new HashMap<>(previousAnalysisModules);
+        Set<String> keys = new HashSet<>();
+        keys.addAll(newAnalysisModules.keySet());
+
+        for (String key : keys) {
+            IAnalysisModule module = newAnalysisModules.remove(key);
+            if (!oldAnalysisModules.containsKey(key)) {
+                /* Once all modules are initialized, automatic modules are executed */
+                if (module != null && module.isAutomatic()) {
+                    status.add(module.schedule());
+                }
+                previousAnalysisModules.put(key, module);
+            } else {
+                oldAnalysisModules.remove(key);
+                fToBeDisposedAnalysisModules.add(module);
```

Comment:
In PR #104 it is possible to replace a data provider factory with a new one. Is there any chance that the new data provider requires a new instance of an analysis module, e.g. that we would need to keep the new one here instead of the old one?

Reply:
OK if I understand correctly your comments, then if a module existed before, the intention is to dispose the old (deferred) and to keep the new (and schedule it if automatic).

Then, if I also understand correctly the code, it will dispose the new (deferred) and keep the old? So it's not doing what's intended?

Should it unconditionally schedule the new and put/replace in previous modules map, and unconditionally remove from old modules map (e.g. doesn´t matter if it's not there and remove does nothing), but if removed then added to deferred to-be-disposed list.

Reply:
Ok, re-read the code to refresh my memory. It's a while that I wrote it. Yes, I could dispose the newly created module right away that is not needed because it already existed. No need to defer them. However, I'll still defer disposal of modules that were removed.